File: sieve.c
#include <stdio.h>

#define N 10000

int sieve(); 
int test_sieve(); 

int sieve() {
	int i;
	int j;
	int a[N+1]; 

	for (i = 2; i <= N; i++) {
		a[i] = 1;
	}

	for (i=2; i <= N/2; i++) {
		for (j=2; j <= N/i; j++) {
			a[i*j] = 0;
		}
	}

	int total = 0; 
	for (i = 1; i <= N; i++) {
		if (a[i]) {
			total += 1;
			printf(" %d ", i); 
		}
	}
	return total;
}

int test_sieve() {
	if (sieve() == 1229) {
		return 1;
	}
	return 0;
}
	
int main() {
	int test_val = sieve();
	if (test_sieve) {
		printf("\nsieve function OK   : got %d\n", test_val); 
	} else {
		printf("\nsieve function FAIL : got %d\n", test_val); 
	}
	return 0;
}

File: sieve.pl
#!/usr/bin/perl

use strict;
use warnings;

our $N = 10000; 
our $NPRIMES = 1229;

sub sieve {
	my $printme = shift; 

	my $i;
	my $j;
	my @a;

	for ($i = 2; $i <= $N; $i++) {
		$a[$i] = 1;
	}

	for ($i = 2; $i <= $N/2; $i++) {
		for ($j = 2; $j <= $N/$i; $j++) {
			$a[$i * $j] = 0;
		}
	}

	my $total = 0;
	for ($i = 1; $i <= $N; $i++) {
		if ($a[$i]) {
			$total += 1;
			if ($printme) {
				print " $i ";
			}
		}
	}
	return $total;
}

sub test_sieve {
	if (sieve(my $printme = 1) == $NPRIMES) {
		return 1;
	} 
	return 0;
}

my $test_val = sieve(my $printme = 0); 
if (test_sieve()) {
	print "sieve function OK   : got $test_val\n"; 
} else {
	print "sieve function FAIL : got $test_val\n"; 
}
	
		




